### PR TITLE
Support to pass externalAssetTag from HTTP header to NotificationSystem.

### PR DESCRIPTION
**Pull request by zzmao**
Add support to pass externalAssetTag from HTTP header to NotificationSystem. 
The String is pass through `BlobProperties`.

### CODE REVIEW

**Comment by zzmao** [comment]
@lightningrob , after a discussion with @cgtz and Matt. We think we will never store this tag to storage node since it may contain PII. 
To avoid confusion in `BlobProperties` where all members are persisted, we moved it to `PutBlobOptions`.  
How do you think of this?

File: ambry-api/src/main/java/com.github.ambry/rest/RestUtils.java [hunk 1]
```diff
@@ -177,6 +177,10 @@
      * The maximum size of the blob that can be uploaded using the URL.
      */
     public static final String MAX_UPLOAD_SIZE = "x-ambry-max-upload-size";
+    /**
+     * An externalAssetTag for this blob.
+     */
+    public static final String EXTERNAL_ASSET_TAG = "x-ambry-external-asset-tag";
```
Comment (lightningrob) [inline ambry-api/src/main/java/com.github.ambry/rest/RestUtils.java]: Ambry client also defines this.  We could move them both to ambryclient-utils module.

**Comment by lightningrob** [comment]
> @lightningrob , after a discussion with @cgtz and Matt. We think we will never store this tag to storage node since it may contain PII.
> To avoid confusion in `BlobProperties` where all members are persisted, we moved it to `PutBlobOptions`.
> How do you think of this?

I guess there's no perfect place to put this tag.  PutBlobOptions doesn't seem like the best fit either, but I don't have a better suggestion.  If we later add any additional parameters, we can combine them into a data structure.

File: ambry-api/src/main/java/com.github.ambry/notification/NotificationSystem.java [hunk 2]
```diff
@@ -34,9 +34,10 @@
    * @param account The {@link Account} for the blob
    * @param container The {@link Container} for the blob
    * @param notificationBlobType The {@link NotificationBlobType} of this blob.
+   * @param externalAssetTag The external asset tag of this blob.
    */
   void onBlobCreated(String blobId, BlobProperties blobProperties, Account account, Container container,
```
Comment (cgtz) [inline ambry-api/src/main/java/com.github.ambry/notification/NotificationSystem.java]: at some point (maybe not this PR), we might want to combine the account, container, notificationBlobType and externalAssetTag arguments into a data structure that gets passed into this method

Comment (zzmao) [reply ambry-api/src/main/java/com.github.ambry/notification/NotificationSystem.java]: @lightningrob , After a discussion with @cgtz , we can actually create a new class to include `Account`, `Container` and `externalAssetTag`. All non-persistent blob properties can be included in this class in the future. Tentatively name it as `NonPersistentBlobProperties`.

Comment (zzmao) [reply ambry-api/src/main/java/com.github.ambry/notification/NotificationSystem.java]: Looks like what I proposed above is not a good way. 
Introducing `NonPersistentBlobProperties`/`BlobNotifications` all requires change to ```Future<String> putBlob(BlobProperties blobProperties, byte[] userMetadata, ReadableStreamChannel channel,
      PutBlobOptions options, Callback<String> callback)```. 
I prefer not to change this elegant interface. 
Will add `externalAssetTag` to `BlobProperties` and comment it as non-persistent property.  @cgtz  @lightningrob

**Comment by zzmao** [comment]
Introduced `externalAssetTag` to `BlobProperties`. Refactor changed a lot files, but main changes are from `BlobProperties.java`, `PutOperation.java`, `PutManagerTest.java`.

File: ambry-api/src/main/java/com.github.ambry/messageformat/BlobProperties.java [hunk 3]
```diff
@@ -137,6 +143,10 @@ public boolean isEncrypted() {
     return isEncrypted;
   }
 
+  public String getExternalAssetTag() {
```
Comment (lightningrob) [inline ambry-api/src/main/java/com.github.ambry/messageformat/BlobProperties.java]: Javadoc (even though existing methods don't all have them)

**Pull request merged**